proposals/0307-add-block-header.md

```diff
@@ -0,0 +1,318 @@
+- Vote timestamps will be removed with Alpenglow.
+
+This SIMD solves these issues by including relevant information in a static
+block header.
+
+## New Terminology
+
+No new terms, but the following definitions are given for clarity:
+
+- Client - The software run by leaders to interface with a solana cluster.
+  (e.g. `agave` or `frankendancer`)
+- Block Producer - The client that produced a given block
+- Scheduler - The system responsible for processing incoming transactions and
+  ordering them for block construction.
+- Forward Error Correction set (FEC set) - A collection of shreds. At a high
+  level, this is a construct that leverages Reed-Solomon encoding to overcome
+  the problem of data loss from packet drops.
+- Shreds - A fixed chunk of encoded raw block data.
+- Entry Batch - An array of entries.
+- Entry - An array of transactions.
+
+## Detailed Design
+
+### Data Layout
+
+Solana blocks are organized in abstraction layers not entirely unlike the
+arrangement of a typical network packet (e.g. MAC -> IP -> TCP -> HTTP). At the
+highest layer a block consists of some number (~100+) FEC sets. A single FEC
+set contains a handful of shreds (~32). Once sufficient shreds are available
+the raw block data is reconstructed and reinterpreted as an array of entry
+batches. Entry batches do not cross shred boundaries.
+
+This SIMD add the following header at the beginning of the raw block data. This
+puts it on the same abstraction layer as serialized entry batch data. Put
+differently, the serialized header will be prepended to the first serialized
+entry batch in the block.
+
+```
+< --       64 bits        -->
++-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+|      block_header_flag    |
++-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+|          version          |
++-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+|       header_length       |
++-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+| block_producer_time_nanos |
++-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+|     block_user_agent      |
+|                           |
+⋮                       +30 ⋮
++-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+|   ... future fields ...   |
++-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+```
+
+- `block_header_flag: u64` will always be zero. The first 8 bytes of an entry
+batch are always a positive number (the number of entries in the batch), so
+this flag allows parsers to differentiate between a normal entry batch and one
+with a header prepended. Though not strictly necessary, this may facilitate
+parsing block data, and allows us to make the header optional if we ever need
+to.
+
+- `version: u64` is a positive integer which changes anytime a change is made to
+the header. The initial version will be 1.
+
+- `header_length: u64` is the length of the rest of the header in bytes (i.e.
+not including the `block_header_flag`, `version`, and `header_length` fields).
+
+- `block_producer_time_nanos: u64` is a nanosecond UNIX timestamp representing
+the time when the block producer became leader and started constructing the
+block.
```

Comment:
Let's be specific that this is the time they started constructing the block. Current description to me had some ambiguity on if it's when I became leader (for my entire allocation) or separately for each block. I think the intent here is a separate timestamp for each block.

Reply:
I can't think of an incentivized reason to lie about this, but it also cannot be verified. Is there a concern about having incorrect values on this?